### PR TITLE
Give better error message when IF Statement has context requirements

### LINKED ISSUE #3146: Inconsistent error message for incompatible return type in nested if / else
This code correctly does not compile. However, the error messages across both branches in the inner if / else are not consistent: -

```fsharp
let x = 1
if x = 1 then true
else
    if x = 2 then "A" // The expression was expected to have type 'bool' but here has type 'string'
    else "B" // All branches of an 'if' expression must have the same return type. This expression was expected to have type 'bool' but here has type 'string'.
```

I would have expected both errors to be the same, using second version (the F#4.1 improved error message).

cc: @forki

### PR DESCRIPTION
**Pull request by forki**
fixes #3146

### CODE REVIEW

File: src/fsharp/FSComp.txt [hunk 1]
```diff
@@ -17,6 +17,7 @@ undefinedNameTypeParameter,"The type parameter %s is not defined."
 undefinedNamePatternDiscriminator,"The pattern discriminator '%s' is not defined."
 replaceWithSuggestion,"Replace with '%s'"
 missingElseBranch,"The 'if' expression is missing an 'else' branch. The 'then' branch has type '%s'. Because 'if' is an expression, and not a statement, add an 'else' branch which returns a value of the same type."
+ifExpression,"The 'if' expression needs to return type '%s' in order to satisfy context type requirements. It currently returns type '%s'."
```
Comment (saul) [inline src/fsharp/FSComp.txt]: 'Returns' may be a bit of a misnomer here - the expression itself has a type. Maybe just: "The 'if' expressions needs to have type '%s' in order to satisfy context type requirements. It currently has type '%s'."

Also out of interest what is a 'context type requirement'?

Comment (forki) [inline src/fsharp/FSComp.txt]: mmm if you have to ask then it's probably not good. 
I wanted to tell that the type requirement comes from the context of the expression - not from the expression itself.

Comment (forki) [inline src/fsharp/FSComp.txt]: @saul @isaacabraham how could we express this better?

**Comment by forki** [comment]
Agent went offline during the build - sigh

File: src/fsharp/FSComp.txt [hunk 2]
```diff
@@ -17,6 +17,7 @@ undefinedNameTypeParameter,"The type parameter %s is not defined."
 undefinedNamePatternDiscriminator,"The pattern discriminator '%s' is not defined."
 replaceWithSuggestion,"Replace with '%s'"
 missingElseBranch,"The 'if' expression is missing an 'else' branch. The 'then' branch has type '%s'. Because 'if' is an expression, and not a statement, add an 'else' branch which returns a value of the same type."
+ifExpression,"The 'if' expression needs to return type '%s' in order to satisfy context type requirements. It currently has type '%s'."
```
Comment (saul) [inline src/fsharp/FSComp.txt]: "needs to return type" -> "must be of type"

File: src/fsharp/FSComp.txt [hunk 3]
```diff
@@ -17,6 +17,7 @@ undefinedNameTypeParameter,"The type parameter %s is not defined."
 undefinedNamePatternDiscriminator,"The pattern discriminator '%s' is not defined."
 replaceWithSuggestion,"Replace with '%s'"
 missingElseBranch,"The 'if' expression is missing an 'else' branch. The 'then' branch has type '%s'. Because 'if' is an expression, and not a statement, add an 'else' branch which returns a value of the same type."
```
Comment (saul) [inline src/fsharp/FSComp.txt]: "returns a value of the same type" -> "has a value of the same type"

File: src/fsharp/FSComp.txt [hunk 4]
```diff
@@ -17,6 +17,7 @@ undefinedNameTypeParameter,"The type parameter %s is not defined."
...
+ifExpression,"The 'if' expression needs to return type '%s' in order to satisfy context type requirements. It currently has type '%s'."
 elseBranchHasWrongType,"All branches of an 'if' expression must return the same type. This expression was expected to have type '%s' but here has type '%s'."
```
Comment (saul) [inline src/fsharp/FSComp.txt]: "must return the same type" -> "must be of the same type"

File: src/fsharp/FSComp.txt (hunk 2 above)
Comment (KevinRansom) [inline src/fsharp/FSComp.txt]: you can lose the 'in order' as well.

File: src/fsharp/FSComp.txt (hunk 3 above)
Comment (forki) [inline src/fsharp/FSComp.txt]: fixed with the other PR

**Comment by forki** [comment]
fixed wording

**Comment by KevinRansom** [comment]
@forki thanks for this.

Kevin

**Pull request merged**